src/history/hs.c

```diff
@@ -838,19 +824,50 @@ __wt_hs_insert_updates(WT_SESSION_IMPL *session, WT_PAGE *page, WT_MULTI *multi)
             if (prev_upd != NULL && prev_upd->start_ts < upd->start_ts)
                 enable_reverse_modify = false;
 
-            /* Find the first update without timestamp. */
-            if (first_non_ts_upd == NULL && upd->start_ts == WT_TS_NONE) {
-                first_non_ts_upd = upd;
-            } else if (first_non_ts_upd != NULL && upd->start_ts != WT_TS_NONE) {
+            if (first_non_ts_upd == NULL) {
+                /* Find the first update without timestamp. */
+                if (upd->start_ts == WT_TS_NONE)
+                    first_non_ts_upd = upd;
+                else if (min_insert_ts < upd->start_ts) {
+                    /*
+                     * Only fix out of order timestamp if we haven't seen an update without
+                     * timestamp.
+                     *
+                     * Resolved prepared updates will lose their durable timestamp here. This is a
+                     * wrinkle in our handling of out-of-order updates.
+                     */
+                    if (upd->start_ts != upd->durable_ts) {
+                        WT_ASSERT(session, min_insert_ts < upd->durable_ts);
+                        WT_STAT_CONN_INCR(session, cache_hs_order_lose_durable_timestamp);
+                    }
+                    __wt_verbose(session, WT_VERB_TIMESTAMP,
+                      "fixing out-of-order updates during insertion; start_ts=%s, "
+                      "durable_start_ts=%s, "
+                      "min_insert_ts=%s",
+                      __wt_timestamp_to_string(upd->start_ts, ts_string[0]),
+                      __wt_timestamp_to_string(upd->durable_ts, ts_string[1]),
+                      __wt_timestamp_to_string(min_insert_ts, ts_string[2]));
+                    upd->start_ts = upd->durable_ts = min_insert_ts;
```

Comment:
if we are changing durable_ts, don't we change the version stability i.e. versions newer than stable timestamp might become stable, if we move the durable_ts from greater than stable timestamp to lesser than the stable timestamp

Reply:
If we need to change the timestamp, that means there is a version with the smaller timestamp newer than us. That update should already be stable.

Reply:
It may be a problem but I won't change it in this PR.